[WebGPUSwift] replace all other Metal specific functions in Command Encoder

nmahendru opened the pull request:
#### 83b8e1d37fa1c1c39f87f84d5abb423e1930fc3f
<pre>
[WebGPUSwift] replace all other Metal specific functions in Command Encoder
<a href="https://rdar.apple.com/144317197">rdar://144317197</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287165">https://bugs.webkit.org/show_bug.cgi?id=287165</a>

Reviewed by Mike Wyrzykowski.

1-to-1 replacement of the below:

1. CommandEncoder::createSimplePso
2. CommandEncoder::runClearEncoder
3. CommandEncoder::clearTextureIfNeeded
4. CommandEncoder::finish

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.swift:
* Source/WebGPU/WebGPU/CommandBuffer.h:
(refCommandBuffer):
(derefCommandBuffer):
* Source/WebGPU/WebGPU/CommandEncoder.h:
(WebGPU::CommandEncoder::create):
(WebGPU::CommandEncoder::setEncoderState):
(WebGPU::CommandEncoder::getEncoderState):
(WebGPU::CommandEncoder::encoderStateNameWrapper):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::waitForCommandBufferCompletion):
(WebGPU::CommandEncoder::encoderIsCurrent const):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_runClearEncoder_thunk(_:attachmentsToClear:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)):
(CommandEncoder_clearTextureIfNeeded_thunk(_:destination:slice:)):
(CommandEncoder_finish_thunk(_:descriptor:)):
(WebGPU.validateFinishError):
(WebGPU.finish(_:)):
(WebGPU.clearTextureIfNeeded(_:slice:)):
(WebGPU.clearTextureIfNeeded(_:slice:device:blitCommandEncoder:)):
(WebGPU.clearTextureIfNeeded(_:_:_:_:_:)):
(WebGPU.sourceBytesPerRow):
(WebGPU.didOverflow):
(WebGPU.checkedBytesPerImage):
(WebGPU.sourceSize):
(WebGPU.sourceBytesPerImage):
(WebGPU.options):
(WebGPU.runClearEncoder(_:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)):
(WebGPU.errorValidatingCopyBufferToBuffer(_:sourceOffset:destination:destinationOffset:size:)):
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingCopyTextureToBuffer(_:destination:copySize:)):
(WebGPU.errorValidatingCopyBufferToTexture(_:destination:copySize:)):
(WebGPU.loadAction(_:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.copyTextureToBuffer(_:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::create):
(WebGPU::Device::Device):
* Source/WebGPU/WebGPU/Internal/Logging.cpp: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
* Source/WebGPU/WebGPU/Internal/Logging.h: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::logString):
(WebGPU_Internal::convertWTFStringToNSString):
(WebGPU_Internal::commandBufferThreadSafeWeakPtr):
(roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong): Deleted.
(roundUpToMultipleOfNonPowerOfTwoUInt32UInt32): Deleted.
(checkedDifferenceSizeT): Deleted.
(isValidToUseWithTextureViewCommandEncoder): Deleted.
(isValidToUseWithQuerySetCommandEncoder): Deleted.
(isValidToUseWithBufferCommandEncoder): Deleted.
(isValidToUseWithTextureCommandEncoder): Deleted.
(clampDouble): Deleted.
(areBuffersEqual): Deleted.
* Source/WebGPU/WebGPU/SwiftCXXThunk.h:

Canonical link: <a href="https://commits.webkit.org/290231@main">https://commits.webkit.org/290231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d02c6605974a4679812e7194a07306c970cb68e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89346 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8871 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94332 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40107 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9258 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68834 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26503 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7105 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49195 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/6851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39214 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77207 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96161 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16526 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77707 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77008 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21432 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9677 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14010 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16540 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16281 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18062 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -233,7 +233,9 @@ class Device : public WGPUDeviceImpl, public ThreadSafeRefCountedAndCanMakeThrea
     };
-
+private PUBLIC_IN_WEBGPU_SWIFT:
+    id<MTLFunction> m_function;
```
Can we rename to `m_nopVertexFunction` or similar please? When it was a function local variable the context was more obvious than its new position as a member variable.

mwyrzykowski commented on Source/WebGPU/WebGPU/Internal/Logging.h:
```diff
@@ -0,0 +1,52 @@
+
+#pragma once
+
+#include "ExportMacros.h"
```
oh this is awesome, thank you! Now we can finally migrate all the WTFLogAlways logging as tracked by https://bugs.webkit.org/show_bug.cgi?id=250442 and improve logging in general as tracked by https://bugs.webkit.org/show_bug.cgi?id=272936

nmahendru replied:
I can rework(in a follow on) and move this to WebGPU from WebGPU_Internal(I intentionally put it there as I don't want to touch WebGPU unless I have too.)
And yea, we should get rid of WTFLogAlways in WebGPU.

mwyrzykowski commented on Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
```diff
@@ -51,28 +53,39 @@ using WTFRangeSizeT = WTF::Range<size_t>;
 __attribute__((used)) static const auto stdDynamicExtent = std::dynamic_extent;
 
+// FIXME: rdar://140819194
+constexpr unsigned long int WGPU_COPY_STRIDE_UNDEFINED_ = WGPU_COPY_STRIDE_UNDEFINED;
```
outside the scope of this PR, but we could also just make `WGPU_COPY_STRIDE_UNDEFINED` a constexpr in WebGPU.h now that we limit the header to C++

nmahendru replied:
This hack is only needed for swift, so isn't it better living here ?

mwyrzykowski replied:
that is fine with me 👍

nmahendru committed:
[WebGPUSwift] replace all other Metal specific functions in Command Encoder
rdar://144317197
https://bugs.webkit.org/show_bug.cgi?id=287165

Reviewed by Mike Wyrzykowski.

1-to-1 replacement of the below:

1. CommandEncoder::createSimplePso
2. CommandEncoder::runClearEncoder
3. CommandEncoder::clearTextureIfNeeded
4. CommandEncoder::finish

* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.swift:
* Source/WebGPU/WebGPU/CommandBuffer.h:
(refCommandBuffer):
(derefCommandBuffer):
* Source/WebGPU/WebGPU/CommandEncoder.h:
(WebGPU::CommandEncoder::create):
(WebGPU::CommandEncoder::setEncoderState):
(WebGPU::CommandEncoder::getEncoderState):
(WebGPU::CommandEncoder::encoderStateNameWrapper):
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::waitForCommandBufferCompletion):
(WebGPU::CommandEncoder::encoderIsCurrent const):
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(CommandEncoder_runClearEncoder_thunk(_:attachmentsToClear:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)):
(CommandEncoder_clearTextureIfNeeded_thunk(_:destination:slice:)):
(CommandEncoder_finish_thunk(_:descriptor:)):
(WebGPU.validateFinishError):
(WebGPU.finish(_:)):
(WebGPU.clearTextureIfNeeded(_:slice:)):
(WebGPU.clearTextureIfNeeded(_:slice:device:blitCommandEncoder:)):
(WebGPU.clearTextureIfNeeded(_:_:_:_:_:)):
(WebGPU.sourceBytesPerRow):
(WebGPU.didOverflow):
(WebGPU.checkedBytesPerImage):
(WebGPU.sourceSize):
(WebGPU.sourceBytesPerImage):
(WebGPU.options):
(WebGPU.runClearEncoder(_:depthStencilAttachmentToClear:depthAttachmentToClear:stencilAttachmentToClear:depthClearValue:stencilClearValue:existingEncoder:)):
(WebGPU.errorValidatingCopyBufferToBuffer(_:sourceOffset:destination:destinationOffset:size:)):
(WebGPU.errorValidatingCopyTextureToTexture(_:destination:copySize:)):
(WebGPU.errorValidatingCopyTextureToBuffer(_:destination:copySize:)):
(WebGPU.errorValidatingCopyBufferToTexture(_:destination:copySize:)):
(WebGPU.loadAction(_:)):
(WebGPU.beginRenderPass(_:)):
(WebGPU.copyTextureToBuffer(_:destination:copySize:)):
(WebGPU.copyBufferToTexture(_:destination:copySize:)):
(WebGPU.copyTextureToTexture(_:destination:copySize:)):
* Source/WebGPU/WebGPU/CommandsMixin.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::create):
(WebGPU::Device::Device):
* Source/WebGPU/WebGPU/Internal/Logging.cpp: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
* Source/WebGPU/WebGPU/Internal/Logging.h: Copied from Source/WebGPU/WebGPU/CommandsMixin.h.
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::logString):
(WebGPU_Internal::convertWTFStringToNSString):
(WebGPU_Internal::commandBufferThreadSafeWeakPtr):
(roundUpToMultipleOfNonPowerOfTwoCheckedUInt32UnsignedLong): Deleted.
(roundUpToMultipleOfNonPowerOfTwoUInt32UInt32): Deleted.
(checkedDifferenceSizeT): Deleted.
(isValidToUseWithTextureViewCommandEncoder): Deleted.
(isValidToUseWithQuerySetCommandEncoder): Deleted.
(isValidToUseWithBufferCommandEncoder): Deleted.
(isValidToUseWithTextureCommandEncoder): Deleted.
(clampDouble): Deleted.
(areBuffersEqual): Deleted.
* Source/WebGPU/WebGPU/SwiftCXXThunk.h:

Canonical link: https://commits.webkit.org/290231@main
